#aMoF - aptamers Motif Finder
###also known as alec's Motif Finder
====
aMoF is a collection of Python scripts developed to to provide a full featured 'motif finder' tools focused on the analysis of APTAMERS selection results.
Classical motifs (or kmers) findinding tools take as input long bio-sequences like entire genomes or sets of genes.  aMoF instead, focuses on sets of short sequences that are the tipical results of an aptamer selection experiment.

##Some background information
Aptamers are short sequence of nucleotides or aminoacids (DNA, RNA, peptides, circular peptides and small proteins) that bind to a specific target molecule.
They can be used for both basic research and clinical purposes as macromolecular drugs or as "atibody-like-particles" for molecular detection.
Aptamers are naturalli present in living organisms where they often act as switch to regulate many different pathways.
In molecular biology, there are two main methods to select aptamers: SELEX and Phage Display. Both methods rely on the analysis of sequences to discriminate specifically enriched aptamers from background noise.

To know more about aptamers here are some links ;)
* http://en.wikipedia.org/wiki/Aptamer
* http://en.wikipedia.org/wiki/Riboswitch
* http://en.wikipedia.org/wiki/Phage_display
* http://en.wikipedia.org/wiki/Systematic_Evolution_of_Ligands_by_Exponential_Enrichment



##How to use aMoF
The core of this repository is aMoF.py. It contains the "motifs finder" code and the specifications for input and output files.


                                       
